### PR TITLE
WIP: implement db for artifacts

### PR DESCRIPTION
**Pull request by CJ-Wright**

### CODE REVIEW

File: libcflib/db.xsh [hunk 1]
```diff
@@ -0,0 +1,82 @@
+"""Database for accessing graph information"""
+import os
+
+import zict
+
+from .model import Artifact
```
Comment (scopatz) [inline libcflib/db.xsh]: Please use full, explicit import, ie `from libcflib.model import Artifact`

File: libcflib/db.xsh [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+class DB:
+    """A logging object for fixie that stores information in line-oriented JSON
+    format.
+    """
+
+    __inst = None
+
+    def __new__(cls):
+        # make the logger a singleton
```
Comment (scopatz) [inline libcflib/db.xsh]: This is not a logger :)

File: libcflib/db.xsh [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        Caching forked from Streamz
+        Copyright (c) 2017, Continuum Analytics, Inc. and contributors
+        All rights reserved.
+
+        """
+        if os.path.exists($LIBCFGRAPH_DIR):
+            git pull $LIBCFGRAPH_URL master
+        else:
+            git clone $LIBCFGRAPH_URL $LIBCFGRAPH_DIR
```
Comment (scopatz) [inline libcflib/db.xsh]: Shouldn't this be a shallow clone?

**Comment by CJ-Wright** [comment]
@justcalamari

File: libcflib/db.xsh [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    __inst = None
+
+    def __new__(cls):
+        # make the logger a singleton
+        if DB.__inst is None:
+            DB.__inst = object.__new__(cls)
+        return DB.__inst
+
+    def __init__(self, history=100):
```
Comment (scopatz) [inline libcflib/db.xsh]: I think history should be called something like `cache_size` and should default to 1 Million

File: libcflib/db.xsh [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        Caching forked from Streamz
+        Copyright (c) 2017, Continuum Analytics, Inc. and contributors
+        All rights reserved.
+
+        """
+        if os.path.exists($LIBCFGRAPH_DIR):
+            git pull $LIBCFGRAPH_URL master
+        else:
+            git clone $LIBCFGRAPH_URL $LIBCFGRAPH_DIR
+        self.cache_dict = {}
```
Comment (scopatz) [inline libcflib/db.xsh]: Probably best not to put type names in variable names.  I would just call this `cache`

File: libcflib/db.xsh [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    __inst = None
+
+    def __new__(cls):
+        # make the db a singleton
+        if DB.__inst is None:
+            DB.__inst = object.__new__(cls)
+        return DB.__inst
+
+    def __init__(self, cache_size=1e6):
```
Comment (scopatz) [inline libcflib/db.xsh]: `1e6` is a float not an int ;)

File: libcflib/db.xsh [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        else:
+            git clone $LIBCFGRAPH_URL $LIBCFGRAPH_DIR --depth 1
+        self.cache = {}
+        self.lru = zict.LRU(cache_size, self.cache)
+        self.times = {}
+
+    def _build_whoosh(self):
+        self.idx = 'whoosh'
+
+
```
Comment (scopatz) [inline libcflib/db.xsh]: There should be only one line between methods

File: libcflib/db.xsh [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                if result not in self.cache:
+                    data = self.get_data(result)
+                    self.cache[result] = data
+                    # Cache the time so we can timeout the record
+                    self.times[result] = time.time()
+                else:
+                    data = self.cache[results]
+                yield data
+
+
```
Comment (scopatz) [inline libcflib/db.xsh]: only one line

File: libcflib/db.xsh (hunk 6 above)
Comment (CJ-Wright) [reply libcflib/db.xsh]: ugh

**Pull request merged**